pkg/operation/botanist/controlplane/etcd/bootstrap_test.go

```diff
@@ -0,0 +1,912 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcd_test
+
+import (
+	"context"
+	"fmt"
+
+	mockclient "github.com/gardener/gardener/pkg/mock/controller-runtime/client"
+	. "github.com/gardener/gardener/pkg/operation/botanist/controlplane/etcd"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	. "github.com/gardener/gardener/pkg/utils/test/matchers"
+
+	resourcesv1alpha1 "github.com/gardener/gardener-resource-manager/pkg/apis/resources/v1alpha1"
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("Etcd", func() {
+	var (
+		ctrl *gomock.Controller
+		c    *mockclient.MockClient
+
+		ctx       = context.TODO()
+		fakeErr   = fmt.Errorf("fake error")
+		namespace = "shoot--foo--bar"
+	)
+
+	BeforeEach(func() {
+		ctrl = gomock.NewController(GinkgoT())
+		c = mockclient.NewMockClient(ctrl)
+	})
+
+	AfterEach(func() {
+		ctrl.Finish()
+	})
+
+	Describe("#BootstrapSeed", func() {
+		var (
+			seedVersion               = "1.12.0"
+			etcdDruidImage            = "etcd/druid:1.2.3"
+			imageVectorOverwriteEmpty *string
+			imageVectorOverwriteFull  = pointer.StringPtr("some overwrite")
+
+			managedResourceName       = "etcd-druid"
+			managedResourceSecretName = "managedresource-" + managedResourceName
+
+			serviceAccountYAML = `apiVersion: v1
+kind: ServiceAccount
```

Comment:
I do not have anything against these yaml definitions per-se, it is just that in other component tests (e.g kube-controller manager), resources are defined with their concrete types.
Do you think it makes sense to have one uniform approach across components ?

Reply:
You are mixing up two things: The bootstrapping resources are not deployed directly to the seed's API server but indirectly via `ManagedResource`s (which work with YAML definitions). The shoot control plane specific resources are deployed directly to the API server, hence, we can also in the test work with the Golang types. Still, for the bootstrapping we are using Golang types as well for the real implementation (see `bootstrap.go`) and marshal them later to YAML when constructing the `ManagedResource`.